### DISCUSSION TITLE
osxphotos saved me!

### DISCUSSION THREAD

**Discussion by dmd** [RhetTbull/osxphotos#1160] [General]
I was having a weird issue where some of my 'year' views didn't show up. I held down option-command and restarted Photos and rebuilt the library. That fixed it, **but...** mysteriously, I noticed that some of my tags _vanished_.  I have tens of thousands of photos and I am extremely careful about my tags, so this was terrifying to me.

Thankfully, just a few days ago I'd happened to run `osxphotos dump`. So, armed with that as a 'before', and a new dump as an 'after', I was able to determine not only what tags were missing, but even which photos they were missing from.  A little bit of bash scripting later (`sort`, `uniq`, `comm`, `csvcut`) and I had something I could write a loop around and add all the tags back in!

`osxphotos dump` going forward will be part of my daily backup script!

**Comment by RhetTbull** [comment]
Great! Glad to hear it! If you use the `--json` option to `dump` or `query` you could save some of the bash scripting (and also get a more complete backup) by getting the output in JSON format which can be manipulated with tools like `jq` or just plain python. The CSV output from both `dump` and `query` dumps most but not all metadata associated with each asset. The `--json` version dumps the everything osxphotos knows in a form that can be reconstructed into a dictionary data structure.

**Comment by RhetTbull** [comment]
By the way, I plan to eventually deprecate `dump` in favor of `query` which can do everything `dump` does and much more. (#793)

**Comment by RhetTbull** [comment]
Yet one more way to backup metadata would be to use [osxphotos sync](https://rhettbull.github.io/osxphotos/cli.html#osxphotos-sync) which is designed for syncing metadata between two different libraries but can be used on a single library.  You could do this to create a backup of all the metadata:

`osxphotos sync --export backup.db`

Then if you wanted to restore tags/keywords at some point in the future, you could do this:

`osxphotos sync --import backup.db --merge keywords` 

Or to overwrite and not merge,

`osxphotos sync --import backup.db --set keywords`

Edit: there is one important difference in how this works to be aware of. The `sync` feature is designed to work between two different libraries where the unique ID (`UUID`) of the photo in the Photos database is not the same between the libraries. Instead, `sync` uses a fingerprint of each photo (a hash effectively) which is guaranteed to be unique and reproducible for a given image (the same algorithm Photos uses to match duplicates).  This means that if you do have true duplicates in the library but with different metadata, their metadata will be merged on export and thus on import, the duplicates will get the same metadata applied even if they weren't that way originally.

**Comment by RhetTbull** [comment]
Another idea I'm posting here in case others have similar issues in the future. If you regularly use `osxphotos export` then a copy of the most recent metadata is saved in the export database (`.osxphotos_export.db` in the export directory).  The metadata is in the `photoinfo` table as a JSON dictionary with primary key of `UUID` of the photo in question. You could always recover the metadata state from the last export in this case. Not as a good as a regular dump like you're doing but good to know if Photos was corrupted but you had recently exported.

**Comment by dmd** [comment]
I can think of times I'd want all of these, so...

```bash
basename=~/Dropbox/bk/photos-dump/photos-dump-$(date -I)
osxphotos query > $basename.csv
gzip $basename.csv
osxphotos query --json > $basename.json
gzip $basename.json
osxphotos sync --export $basename.db
gzip $basename.db
```

:)
